src/optimizer/expr_simplification.ts

```diff
@@ -0,0 +1,713 @@
+     *   the fields, constants and methods in the ast property are NOT coalesced. This means, for example,
+     *   that the methods in a TypeDescription's ast of a contract will be methods that are actually
+     *   declared in the contract and not in some trait that the contract inherits from.
+     *
+     * The above means that we will need to process the properties in TypeDescription first,
+     * and then use those properties to build the AST (carefully ensuring that only fields, constants and methods
+     * that were in the original AST, remain in the new AST).
+     */
+    const newTypes: Map<string, TypeDescription> = new Map();
+
+    for (const t of getAllTypes(ctx)) {
+        let newInitializer: InitDescription | null = null;
+
+        // Process init
+        if (t.init) {
+            const statementsResult = process_statements(
+                t.init.ast.statements,
+                ctx,
+                interpreter,
+            );
+            const newStatements = statementsResult.stmts;
+            ctx = statementsResult.ctx;
+            const newInitCode = cloneAstNode({
+                ...t.init.ast,
+                statements: newStatements,
+            });
+            newInitializer = { ...t.init, ast: newInitCode };
+        }
+
+        // Process constants
+        const newConstants: ConstantDescription[] = [];
+
+        // This map will be used to quickly recover the new definitions when
+        // building the AST later
+        const newConstantsMap: Map<string, ConstantDescription> = new Map();
+
+        for (const c of t.constants) {
+            if (c.ast.kind === "constant_def") {
+                const expressionResult = process_expression(
+                    c.ast.initializer,
+                    ctx,
+                    interpreter,
+                );
+                const newInitializer = expressionResult.expr;
+                ctx = expressionResult.ctx;
+                const newConstantCode = cloneAstNode({
+                    ...c.ast,
+                    initializer: newInitializer,
+                });
+                const newConstantDescription = { ...c, ast: newConstantCode };
+                newConstants.push(newConstantDescription);
+                newConstantsMap.set(c.name, newConstantDescription);
+            } else {
+                // The rest of kinds do not have explicit Tact expressions, so just copy the current description
+                newConstants.push(c);
+                newConstantsMap.set(c.name, c);
+            }
+        }
+
+        // Process fields
+        const newFields: FieldDescription[] = [];
+
+        // This map will be used to quickly recover the new definitions when
+        // building the AST later
+        const newFieldsMap: Map<string, FieldDescription> = new Map();
+
+        for (const f of t.fields) {
+            if (f.ast.initializer !== null) {
```

Comment:
Bodies of these loops look very similar. Also if there is a new place in AST that needs to be optimized, there are no failsafes to ensure it will be handled in optimizer.